### PR TITLE
Add cache BBE

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject

### CODE REVIEW

File: new-examples/cache-invalidation/cache_invalidation.bal [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import ballerina/cache;
+import ballerina/io;
+
+public function main() returns error? {
+
+    // This creates a new cache with custom configuration.
```
Comment (daneshk) [inline new-examples/cache-invalidation/cache_invalidation.bal:6]:
```suggestion
    // This creates a new cache with advanced invalidation configuration.
```

File: new-examples/index.json [hunk 2]
```diff
@@ -64,5 +64,20 @@
...
             }
         ]
+    },
+    {
+        "title": "Cache",
+        "column": 0,
+        "category": "Language concepts",
+        "samples": [
+            {
+                "name": "Cache Basic Operations",
```
Comment (daneshk) [inline new-examples/index.json:74]:
```suggestion
                "name": "Cache Basics",
```

File: new-examples/cache-basics/cache_basics.bal [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import ballerina/cache;
+import ballerina/io;
+
+public function main() returns error? {
+    // This creates a new cache instance with default configurations.
```
Comment (praneesha) [inline new-examples/cache-basics/cache_basics.bal:5]:
```suggestion
    // This creates a new cache instance with the default configurations.
```

File: new-examples/cache-basics/cache_basics.metatags [hunk 4]
```diff
@@ -0,0 +1,2 @@
+description: This BBE shows how to perform an in-memory caching basic operations with the "Least Recently Used" algorithm in Ballerina.
```
Comment (praneesha) [inline new-examples/cache-basics/cache_basics.metatags:1]:
```suggestion
description: This BBE shows how to perform basic in-memory caching operations with the "Least Recently Used" algorithm in Ballerina.
```

File: new-examples/cache-basics/tests/cache_test.bal [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import ballerina/test;
+
+string[] outputs = [];
+
+// This is the mock function which will replace the real function
```
Comment (praneesha) [inline new-examples/cache-basics/tests/cache_test.bal:5]:
```suggestion
// This is the mock function, which will replace the real function.
```

File: new-examples/cache-basics/tests/cache_test.bal [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    } else {
+        outputs.push(value.toString());
+    }
+}
+
+@test:Config{}
+function testFunc() returns error? {
+    test:when(mock_printLn).call("mockPrint");
+
+    // Invoking the main function
```
Comment (praneesha) [inline new-examples/cache-basics/tests/cache_test.bal:25]:
```suggestion
    // Invoking the main function.
```

File: new-examples/cache-invalidation/cache_invalidation.bal [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import ballerina/cache;
+import ballerina/io;
+
+public function main() returns error? {
+
+    // This creates a new cache with advanced configuration.
```
Comment (praneesha) [inline new-examples/cache-invalidation/cache_invalidation.bal:6]:
```suggestion
    // This creates a new cache with the advanced configuration.
```

File: new-examples/cache-invalidation/cache_invalidation.bal [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    cache:Cache cache = new ({
+        // The maximum size of the cache 10.
```
Comment (praneesha) [inline new-examples/cache-invalidation/cache_invalidation.bal:8]:
```suggestion
        // The maximum size of the cache is 10.
```

File: new-examples/cache-invalidation/cache_invalidation.bal [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        // time of eviction 10*0.2=2 entries get removed from the cache.
+        evictionFactor: 0.2,
+        // The default max age of the cache entry is set to 2 seconds.
+        defaultMaxAge: 2,
+        // The cache cleanup task runs every 3 seconds and clears all
+        // the expired entries.
+        cleanupInterval: 3
+    });
+
+    // Adds new entries to the cache.
```
Comment (praneesha) [inline new-examples/cache-invalidation/cache_invalidation.bal:20]:
```suggestion
    // Adds the new entries to the cache.
```

File: new-examples/cache-invalidation/cache_invalidation.bal [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    check cache.put("key3", "value3", 3600);
+
+    // Gets the keys of the cache entries.
+    string[] keys = cache.keys();
+    io:println("The existing keys in the cache: ", keys);
+
+    // Discards the given cache entry.
+    _ = check cache.invalidate("key2");
+
+    // Gets the keys of the cache entries
```
Comment (praneesha) [inline new-examples/cache-invalidation/cache_invalidation.bal:33]:
```suggestion
    // Gets the keys of the cache entries.
```

File: new-examples/cache-invalidation/cache_invalidation.bal [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    io:println("The existing keys in after invalidating a given key: ", cache.keys());
+
+    // Discards all the cache entries of the cache.
+    _ = check cache.invalidateAll();
+
+    // Gets the keys of the cache entries, after invalidated all the keys.
```
Comment (praneesha) [inline new-examples/cache-invalidation/cache_invalidation.bal:39]:
```suggestion
    // Gets the keys of the cache entries after all the keys are invalidated.
```

File: new-examples/cache-invalidation/cache_invalidation.bal [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    io:println("The keys in after invalidating all the keys: ", cache.keys());
```
Comment (praneesha) [inline new-examples/cache-invalidation/cache_invalidation.bal:40]:
```suggestion
    io:println("The keys after invalidating all the keys: ", cache.keys());
```

File: new-examples/cache-invalidation/cache_invalidation.description [hunk 13]
```diff
@@ -0,0 +1,4 @@
+// The Ballerina Cache API provides an in-memory cache implementation by default with a
```
Comment (praneesha) [inline new-examples/cache-invalidation/cache_invalidation.description:1]:
```suggestion
// The Ballerina Cache API provides an in-memory cache implementation by default with the
```

File: new-examples/cache-invalidation/tests/cache_test.bal [hunk 14]
```diff
@@ -0,0 +1,40 @@
+import ballerina/test;
+
+string[] outputs = [];
+
+// This is the mock function which will replace the real function
```
Comment (praneesha) [inline new-examples/cache-invalidation/tests/cache_test.bal:5]:
```suggestion
// This is the mock function, which will replace the real function.
```

File: new-examples/cache-invalidation/tests/cache_test.bal [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    } else {
+        outputs.push(value.toString());
+    }
+}
+
+@test:Config{}
+function testFunc() returns error? {
+    test:when(mock_printLn).call("mockPrint");
+
+    // Invoking the main function
```
Comment (praneesha) [inline new-examples/cache-invalidation/tests/cache_test.bal:25]:
```suggestion
    // Invoking the main function.
```

**Comment by praneesha** [comment]
Please update the keywords and descriptions of all the BBEs.

**Comment by kalaiyarasiganeshalingam** [comment]
> Please update the keywords and descriptions of all the BBEs.
I have already updated those. Please check
https://github.com/ballerina-platform/ballerina-distribution/pull/1746/files#diff-322c39bf49ea78c7747633aef32e7a9cde7e68f31fed373092a9e7e6deb196bcR1
https://github.com/ballerina-platform/ballerina-distribution/pull/1746/files#diff-916a149811c4020e159f6f3b9fbda3836ceb79f2fde1e14ffdad8fda9103ecebR1

**Review by daneshk — APPROVED** [review]
LGTM

File: new-examples/cache-invalidation/cache_invalidation.metatags [hunk 16]
```diff
@@ -0,0 +1,2 @@
+description: This BBE shows how to perform an in-memory caching invalidate operation with the "Least Recently Used" algorithm in Ballerina.
+keywords: ballerina, ballerina by examples, bbe, cache, put, hasKey, invalidate, invalidateAll
```
Comment (praneesha) [inline new-examples/cache-invalidation/cache_invalidation.metatags:2]:
```suggestion
keywords: ballerina, ballerina by examples, BBE, cache, put, hasKey, invalidate, invalidateAll
```
Please correct in all occurrences.

**Review by praneesha — APPROVED** [review]

**Pull request merged**